### ISSUE TITLE
Merge ams/next branch

### ISSUE THREAD

**Issue by ams** [LM-3/uhdl#7]
I've been working on cleaning up the code somewhat, adding back the Verilator simulator code, and removing b0rked code.  Plus allowing to build using make instead of Xilinx ISE.  Featurewise this is the same as before, but less code, less jumping around, more documentation and more testbenches.

**Comment by ams** [comment]
FWIW, ams/next is mainly offline and I haven't bothered pushing it at this time.

**Comment by nusgart** [comment]
@ams I'd appreciate if you would push the ams/next branch since I am interested in helping with this, but I don't want to make changes that conflict with yours.

**Comment by ams** [comment]
Fair enough! I’ll see what I can do.  The MMC code really could need a whack indeed.  I have some iCE40s that I was thinking might be an interesting target for this, what’s your take in that?

**Comment by nusgart** [comment]
I think that the iCE40 would be an interesting target and they do look to be large enough, except possibly be the amount of embedded RAM, based on the amount ISE seems to generate, but I think the larger ones should be fine.  However, feasibility-wise, I'm not sure how hard a port would be -- my experience with FPGAs is slightly limited, especially in porting from one vendor to another (I was thinking to another spartan 6 dev board for the immediate future), but I would be interested in helping.  I'll definitely be interested in improving the MMC code, though.

**Comment by ams** [comment]
Great! I'll try and merge my hacks this weekend/today, would like to double check that the code still works on the my Pipistrello.  I should warn that the merge might be a big blob commit, there is a lot of clean up which I think will be useful when trying to get this working on another board.  Tried to use some generic code between modules for the BRAM and other bits, but also restored a bunch of the old test benches and Verilator simulator that existed at some point.  There is also a very early start -- though very incomplete -- to a Chaosnet interface that I started working on -- plan was to use another UART line instead of trying to use a Ethernet PHY since that is easier to get working.

**Comment by ams** [comment]
Also, hopefully you don't mind me using verilog-mode hackery, and https://github.com/ams/hdlmake.mk for building the bitstreams/testbenches.  Though everything works in Xilinx ISE as well, only that you won't get the nice AUTOINST/AUTOARG/... expansions.

**Comment by nusgart** [comment]
Both of those are fine by me.

**Comment by ams** [comment]
There you go.

**Comment by ams** [comment]
@nusgart Did you have any luck getting things running, or did the commit scare you? :-)

**Comment by nusgart** [comment]
@ams I know that it has been a long time, but I am still here.  I've been busy attempting to port this to various FGPAs.  I started trying to port it to the Altera Cyclone 2 and then the Max 10. However, I eventually discovered that Altera's dram IP cores only work for tethered systems unless you buy them, and among other things, this convinced me to try an alternate path.
I then decided to port this to Diligent's Arty A7-35  board because the A7 is more similar to the Spartan 6, but is supported by more modern tools, and is more affordable for the same power.  Also, if you're interested, project x-ray will hopefully enable the same open-source development flow as with the iCE40s.  For now, though, I'm using Vivado 2018.3 to build the project, and I've gotten all the way to successful bitgen.  However, I haven't actually been able to test it on real hardware yet, and there are some timing warnings that might be problematic.  I've also probably broken the build for the LX45 since Vivado and ISE disagree on some of the finer points of Verilog.  The commit history also needs to be cleaned since I didn't properly separate all of the toplevel files at first.
